### PR TITLE
Added the automatic update of feedstocks through PRs.

### PR DESCRIPTION
**Pull request by pelson**
Still very much a work in progress, but the aspiration is to:

**Automatically run conda-smithy on all feedstocks, proposing new PRs for any that are out of date.**

I plan to cron this task, and to push a maximum of one PR per hour (initially) to avoid overloading the CI systems. This will allow us to update conda-smithy, and gradually roll out any administrative type changes to the whole feedstock collection.

### CODE REVIEW

File: scripts/regenerate_feedstock.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        smithy_rerender onto the branch
+        if changes commit
+        push branch to remote
+        if pull request for branch, add a comment
+        else create a pull request
+"""
+
+import os
+feedstocks_dir = os.path.expanduser("~/dev/conda-forge/feedstocks")
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Don't hard-code this quite so badly.

File: scripts/regenerate_feedstock.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+#feedstocks.clone_all('conda-forge', feedstocks_dir)
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Re-enable this. (It does take a while, but is necessary)

File: scripts/regenerate_feedstock.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+#feedstocks.fetch_feedstocks(feedstocks_dir)
+# TODO: What about feedstocks that get removed?
+
+import random
+randomised_feedstocks = list(feedstocks.cloned_feedstocks(feedstocks_dir))
+# Shuffle is in-place. :(
+random.shuffle(randomised_feedstocks)
+
+import git
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Pull imports up

File: scripts/regenerate_feedstock.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+import github
+import conda_smithy.github
+
+# XXX Ensure this is the conda-forge-admin token
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Done later on.

File: scripts/regenerate_feedstock.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+if gh_me.login != 'conda-forge-admin':
+    raise ValueError("The github token isn't that of conda-forge-admin (it's "
+                     "for {}), I'm going to have to bail.".format(gh_me.login))
+
+
+gh_forge = gh.get_organization('conda-forge')
+
+
+def my_repos(gh_user):
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Document why these exist, and link to the opened issue.

File: scripts/regenerate_feedstock.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        my_repos[feedstock.name] = gh_me.get_repo(feedstock.name)
+
+    clone = git.Repo(feedstock.directory)
+    admin_fork = my_repos[feedstock.name]
+    forge_feedstock = forge_repos[feedstock.name]
+
+    # Put an appropriate conda-forge-admin remote in place.
+    if 'conda-forge-admin' in [remote.name for remote in clone.remotes]:
+        clone.delete_remote('conda-forge-admin')
+    clone.create_remote('conda-forge-admin', url=admin_fork.clone_url.replace('https://', 'https://{}@'.format(gh_token)))
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Should this remote be removed when the process is complete. Perhaps so.

File: scripts/regenerate_feedstock.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    if 'feedstock_rerender' in clone.heads:
+        clone.heads.master.checkout()
+        clone.delete_head('feedstock_rerender', '-D')
+    clone.create_head('feedstock_rerender', clone.remotes.upstream.refs.master).set_tracking_branch(clone.remotes.upstream.refs.master)
+
+    # Reset the working tree to a clean state.
+    clone.head.reset(index=True, working_tree=True)
+    clone.heads.feedstock_rerender.checkout()
+
+    import conda_smithy.configure_feedstock
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Pull to top.

File: scripts/regenerate_feedstock.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        diff = commit.diff(clone.remotes['conda-forge-admin'].refs.feedstock_rerender)
+        if not diff:
+            # There were no differences between this and the remote feedstock_rerender, so just continue.
+            print("{} was checked, and whilst there are changes needed, the PR is up-to-date".format(feedstock.name))
+            continue
+
+    clone.remotes['conda-forge-admin'].push('+feedstock_rerender')
+
+    rerender_pulls = list(list_pulls(forge_feedstock, state='open', head='conda-forge-admin:feedstock_rerender'))
+    CR = "\n"
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Delete

File: scripts/regenerate_feedstock.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    if rerender_pulls:
+        pull = rerender_pulls[0]
```
Comment (pelson) [inline scripts/regenerate_feedstock.py]: Verify this is doing what I want by closing a PR and re-running.

**Comment by pelson** [comment]
Problematic feedstocks:

 * https://github.com/conda-forge/arm_pyart-feedstock/pull/3

**Comment by ocefpaf** [comment]
@pelson this is really cool!

**Comment by pelson** [comment]
> @pelson this is really cool!

**Pull request merged**